doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+
+The management of DPU overlay databases within the NPU is orchestrated through existing SONiC database services. The daemon, named "featured," retains the responsibility for initiating, terminating, enabling, and disabling the DPU overlay database services. This interaction is facilitated using the systemctl tool.
+
+To determine the DPU number, the "featured" daemon should leverage the platform API. However, for the sake of implementation simplicity, the DPU number is extracted directly from the platform_env.conf file firstly.
+
+``` shell
+cat /usr/share/sonic/device/$PLATFORM/platform_env.conf
+NUM_DPU=2
+```
+
+To align with the established multi-ASIC design in SONiC, a new field, `"has_per_dpu_scope": "True"``, is introduced in the database feature table within config_db.json. This field plays a crucial role in ensuring that each DPU database instance is initiated within a dedicated database container. This design approach maintains consistency with SONiC's existing architecture while accommodating the specific requirements of DPU overlay databases.
+
+``` json
+# config_db.json
+
+"database": {
+    "auto_restart": "always_enabled",
+    "delayed": "False",
+    "has_global_scope": "True",
+    "has_per_asic_scope": "True",
+    "has_per_dpu_scope": "True", # New field for DPU database service
+    "high_mem_alert": "disabled",
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6380 + DPU ID). Additionally, the Unix domain socket path is structured as /var/run/redisdpu{DPU_ID}.
+
+Here is an example includes two DPU:
+
+``` json
+# DPU0: /var/run/redisdpu2/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6381,
+    "unix_socket_path": "/var/run/redisdpu0/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+#DPU1: /var/run/redisdpu1/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6382,
+    "unix_socket_path": "/var/run/redisdpu1/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+```
+
+There are four new tables introduction for the DPU overlay database:
+
+``` json
+"DPU_APPL_DB": {
+    "id": 15,
+    "separator": ":",
+    "instance": "redis",
+    "format": "proto"
+},
+"DPU_APPL_STATE_DB": {
+    "id": 16,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_STATE_DB": {
+    "id": 17,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_COUNTERS_DB": {
+    "id": 18,
+    "separator": ":",
```

Comment:
should counters also uses |?

Reply:
in case ipv6 stuffs showing up and complicates the splitting logic.

Reply:
Here is just a SONiC convention. https://github.com/sonic-net/sonic-buildimage/blob/ada7c6a72e27a9729628f383f48cd5f75d4da227/dockers/docker-database/database_config.json.j2#L32
I'm worried that there is some hard-code separator `:` in the existing code base.

But you are right, there is some conflict with ipv6. This may happen in APPL_DB also.
@qiluo-msft do you know why we use the colon as the separator?

Reply:
This is as per design for some DBs. With IPv6, we ensure IPv6 is always at the end of key so as to not conflict with expected Seperator